sql: pool some allocations during flow setup

yuzefovich opened the pull request:
**flowinfra: slightly tweak the setup of processors in the flow**

Previously, the last processor ('headProc') would be removed from
f.processors before flow.startInternal to be run in the same goroutine
as the one that is doing the setup. However, this was problematic
because if that processor implements 'Releasable' interface, it will not
be returned to the pool on the flow clean up. Now this is fixed.

Addresses: #42770.

Release note: None

**sql: pool flow allocations**

sql: pool flow allocations

Previously, new structs for rowBasedFlow and vectorizedFlow would be
allocated upon creation. This commit creates pools for both of them.
flowinfra.Releasable interface is moved into execinfra package because
now components from rowflow, rowexec, and colflow packages implement
that.

In order to actually be able to release the flow structs, I needed to
create separate Cleanup methods (which still share most of the logic)
which allows for removal of vectorized memory monitoring logic from
the shared FlowCtx.

Release note: None

yuzefovich committed:
flowinfra: slightly tweak the setup of processors in the flow

Previously, the last processor ('headProc') would be removed from
f.processors before flow.startInternal to be run in the same goroutine
as the one that is doing the setup. However, this was problematic
because if that processor implements 'Releasable' interface, it will not
be returned to the pool on the flow clean up. Now this is fixed.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42809)
<!-- Reviewable:end -->

ajwerner commented on pkg/sql/flowinfra/flow.go:
```diff
@@ -123,13 +123,14 @@ type Flow interface {
 	// isVectorized indicates whether it is a vectorized flow.
 	isVectorized bool
-	// processors contains a subset of the processors in the flow - the ones that
+	// Processors contains a subset of the processors in the flow - the ones that
```
Rather than exporting all of this stuff would it be better to call into FlowBase.Cleanup() from the various flow implementations which embed it and still have those implementations implement a Cleanup method?

yuzefovich replied:
Yes, you're right. At first I thought that cleaning up of memory monitoring infrastructure in `vectorizedFlow`s forces us to actually copy-paste most of the logic, but rearranging the order of things slightly does the job. Done.

ajwerner reviewed (commented):
Thanks for looking into this so rapidly! I’m hopeful it will move a needle on a benchmark but we’ll see 🤷‍♀️

I’m afk for the next few hours but here’s a passing question.

yuzefovich committed:
sql: pool flow allocations

Previously, new structs for rowBasedFlow and vectorizedFlow would be
allocated upon creation. This commit creates pools for both of them.
flowinfra.Releasable interface is moved into execinfra package because
now components from rowflow, rowexec, and colflow packages implement
that.

In order to actually be able to release the flow structs, I needed to
create separate Cleanup methods (which still share most of the logic)
which allows for removal of vectorized memory monitoring logic from
the shared FlowCtx.

Release note: None

ajwerner reviewed (commented):
Thanks for jumping on this so quickly! Looking forward to seeing if it show up in the top-level benchmarks. I can look into the table descriptors. It seems like there ought to be something straight-forward we can do. 

> Reviewed 1 of 1 files at r1, 5 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42809#-:-Luf2tbf9t9Z2V_rwWBV:bcutn1z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colflow/vectorized_flow.go, line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42809#-Luf2IiUAc7T32VQ6L8u:-Luf2IiV8Yjq2mx6nugL:b-pmx231) ([raw file](https://github.com/cockroachdb/cockroach/blob/330c75271fcb2af4e9e63349f55ef9f0c6a1cc6b/pkg/sql/colflow/vectorized_flow.go#L125)):*
> ```Go
> 	f.bufferingMemAccounts = nil
> 	f.bufferingMemMonitors = nil
> 	vectorizedFlowPool.Put(f)
> ```

I'd prefer`*f = vectorizedFlow{}`

Is there a reason while you're holding on to the reference to the `FlowBase`?
___
*[pkg/sql/rowflow/row_based_flow.go, line 399 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42809#-Luf2eOTDrbcMM_vpsjD:-Luf2eOTDrbcMM_vpsjE:b2wm48b) ([raw file](https://github.com/cockroachdb/cockroach/blob/330c75271fcb2af4e9e63349f55ef9f0c6a1cc6b/pkg/sql/rowflow/row_based_flow.go#L399)):*
> ```Go
> // Release releases this rowBasedFlow back to the pool.
> func (f *rowBasedFlow) Release() {
> 	f.localStreams = nil
> ```

I'd prefer `*f = rowBasedFlow{}`.

Is there a reason while you're holding on to the reference to the `FlowBase`?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for pointing out this issue!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42809#-:-Lui-KiR56N4NocX_Zfb:b-fjsitp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jordanlewis)

___
*[pkg/sql/colflow/vectorized_flow.go, line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42809#-Luf2IiUAc7T32VQ6L8u:-Lui-CTVCR67ZgRVAiuB:b-bqyst5) ([raw file](https://github.com/cockroachdb/cockroach/blob/330c75271fcb2af4e9e63349f55ef9f0c6a1cc6b/pkg/sql/colflow/vectorized_flow.go#L125)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'd prefer`*f = vectorizedFlow{}`

Is there a reason while you're holding on to the reference to the `FlowBase`?
</blockquote></details>

No reason, I forgot about such zeroing out, thanks. Done.
___
*[pkg/sql/rowflow/row_based_flow.go, line 399 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42809#-Luf2eOTDrbcMM_vpsjD:-Lui-JVe5qxcMXzO2CAV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/330c75271fcb2af4e9e63349f55ef9f0c6a1cc6b/pkg/sql/rowflow/row_based_flow.go#L399)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'd prefer `*f = rowBasedFlow{}`.

Is there a reason while you're holding on to the reference to the `FlowBase`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42809#-:-Lui2F_u8lOuAyHCeJq3:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+

ajwerner commented:
```
name             old ops/s   new ops/s   delta
KV95-throughput  14.3k ± 2%  14.9k ± 2%  +4.08%  (p=0.008 n=5+5)

name             old ms/s    new ms/s    delta
KV95-P50          9.10 ± 3%   8.90 ± 0%    ~     (p=0.444 n=5+5)
KV95-Avg          2.72 ± 4%   2.54 ± 2%  -6.62%  (p=0.048 n=5+5)
```
Not too shabby! I ran it with a higher concurrency the first time and it didn't show up as well:

```
name             old ops/s   new ops/s   delta
KV95-throughput  15.0k ± 2%  15.6k ± 4%    ~     (p=0.056 n=5+5)

name             old ms/s    new ms/s    delta
KV95-P50          11.5 ± 5%   11.3 ± 7%    ~     (p=0.492 n=5+5)
KV95-Avg          3.40 ± 0%   3.30 ± 0%  -2.94%  (p=0.029 n=4+4)
```

I'm going to chalk it up as a real win.
